dispatcher.go

```diff
@@ -45,70 +52,149 @@ type Dispatcher[T any] interface {
+	// NewFilteredEventStream creates a new event stream that will only receive events that
+	// match the provided condition. Note that the condition will be checked prior to
+	// dispatch, which is more performant than filtering in the event handler itself.
 	NewFilteredEventStream(condition EventCondition[T]) EventStream[T]
 
 	// CloseEventStreams closes all listening event streams and shuts down the dispatcher
 	CloseEventStreams()
 }
 
+//--------------------------------------------------------------------------
+//  EventStream[T] Helpers
+//--------------------------------------------------------------------------
+
+// SyncEvent[T] is an event wrapper which contains the event T payload and a channel to
+// signal when the event is processed.
+type SyncEvent[T any] struct {
+	closed *atomicbool
+	done   chan struct{}
+
+	// Event contains the T event payload that was dispatched.
+	Event T
+}
+
+// Done notifies the event system that the event has been processed by the receiver.
+func (se *SyncEvent[T]) Done() {
+	if !se.closed.CompareAndSet(false, true) {
+		return
+	}
+
+	if se.done != nil {
+		close(se.done)
+	}
+}
+
+// creates a new synchronous event wrapper to pass to the synchronous event stream.
+func newSyncEvent[T any](event T) SyncEvent[T] {
+	return SyncEvent[T]{
+		closed: newAtomicBool(false),
+		done:   make(chan struct{}),
+		Event:  event,
+	}
+}
+
+// creates a new faux sync event, which is used when there exists a synchronous receiver for
+// an event dispatched asynchronously
+func newFauxSyncEvent[T any](event T) SyncEvent[T] {
+	return SyncEvent[T]{
+		closed: newAtomicBool(false),
+		done:   nil,
+		Event:  event,
+	}
+}
+
 //--------------------------------------------------------------------------
 //  Default EventStream[T] Implementation
 //--------------------------------------------------------------------------
 
 // asyncEventStream contains the event stream channel and metadata
 type asyncEventStream[T any] struct {
-	stream    chan T
-	closed    *atomicbool
-	condition EventCondition[T]
+	stream     chan T
+	syncStream chan SyncEvent[T]
+	closed     *atomicbool
+	accessed   *atomicbool
+	sync       *atomicbool
+	condition  EventCondition[T]
 }
 
 // newAsyncEventStream creates a new asynchronous event stream for a listener.
 func newAsyncEventStream[T any](condition EventCondition[T]) *asyncEventStream[T] {
 	return &asyncEventStream[T]{
-		closed:    newAtomicBool(false),
-		stream:    make(chan T),
-		condition: condition,
+		accessed:   newAtomicBool(false),
+		sync:       newAtomicBool(false),
+		closed:     newAtomicBool(false),
+		stream:     make(chan T),
+		syncStream: make(chan SyncEvent[T]),
+		condition:  condition,
 	}
 }
 
 // Stream returns access to the event T channel where events will arrive.
 func (aes *asyncEventStream[T]) Stream() <-chan T {
+	if aes.accessed.CompareAndSet(false, true) {
+		aes.sync.Set(false)
+	}
+
+	if aes.sync.Get() {
+		return nil
+	}
```

Comment:
What does this do?

Reply:
A likely explosion - I didn't make a decision here, but we should probably panic, as I don't think there is a good way forward if this access happens.